addons/delivery/delivery.py

```diff
@@ -48,135 +46,130 @@ def get_price(self, cr, uid, ids, field_name, arg=None, context=None):
         if isinstance(ids, (int,long)):
             ids = [ids]
+        if vals.get('country_ids'):
+            country_ids = [country['id'] for country in self.resolve_2many_commands(cr, uid, 'country_ids', vals.get('country_ids'), context)]
+            for carrier in self.browse(cr, uid, ids, context):
+                if carrier.state_ids:
+                    vals['state_ids'] = [(3, state.id) for state in carrier.state_ids.filtered(lambda x: x.country_id.id not in country_ids)]
         res = super(delivery_carrier, self).write(cr, uid, ids, vals, context=context)
-        self.create_grid_lines(cr, uid, ids, vals, context=context)
+        self.create_price_rules(cr, uid, ids, vals, context=context)
         return res
 
     def create(self, cr, uid, vals, context=None):
+        vals = self.filter_state_country(cr, uid, vals, context)
+        vals = self.create_product_for_delivery(cr, uid, vals, context=context)
         res_id = super(delivery_carrier, self).create(cr, uid, vals, context=context)
-        self.create_grid_lines(cr, uid, [res_id], vals, context=context)
+        self.create_price_rules(cr, uid, [res_id], vals, context=context)
         return res_id
 
+    def onchange_states(self, cr, uid, ids, state_ids, country_ids, context=None):
+        if state_ids:
+            state_country_ids = [state['country_id'][0] for state in self.resolve_2many_commands(cr, uid, 'state_ids', state_ids, ['country_id'], context)]
+            country_ids = [country['id'] for country in self.resolve_2many_commands(cr, uid, 'country_ids', country_ids, context)]
+            return {'value': {'country_ids': [(6, 0, state_country_ids + country_ids)]}}
+        return {}
 
-class delivery_grid(osv.osv):
-    _name = "delivery.grid"
-    _description = "Delivery Grid"
-    _columns = {
-        'name': fields.char('Grid Name', required=True),
-        'sequence': fields.integer('Sequence', required=True, help="Gives the sequence order when displaying a list of delivery grid."),
-        'carrier_id': fields.many2one('delivery.carrier', 'Carrier', required=True, ondelete='cascade'),
-        'country_ids': fields.many2many('res.country', 'delivery_grid_country_rel', 'grid_id', 'country_id', 'Countries'),
-        'state_ids': fields.many2many('res.country.state', 'delivery_grid_state_rel', 'grid_id', 'state_id', 'States'),
-        'zip_from': fields.char('Start Zip', size=12),
-        'zip_to': fields.char('To Zip', size=12),
-        'line_ids': fields.one2many('delivery.grid.line', 'grid_id', 'Grid Line', copy=True),
-        'active': fields.boolean('Active', help="If the active field is set to False, it will allow you to hide the delivery grid without removing it."),
-    }
-    _defaults = {
-        'active': lambda *a: 1,
-        'sequence': lambda *a: 1,
-    }
-    _order = 'sequence'
-
-    def get_price(self, cr, uid, id, order, dt, context=None):
+    def get_price_available(self, cr, uid, id, order, dt, context=None):
```

Comment:
get_price_available method should be process on batch. so change method prototype . use ids instead of id.